### PR TITLE
fix visibility bug

### PR DESCRIPTION
**Pull request by yingjunwu**
The insertion/deletion bugs reported recently are mostly related to the design of GC manager. This PR tries to fundamentally solve this problem by redesigning the GC component.

### CODE REVIEW

**Comment by apavlo** [comment]
This is fantastic. Can you be specific about what bugs you are fixing (i.e., #756, #712, #713, #709)?

**Comment by yingjunwu** [comment]
I can confirm that this PR can fix #709 and #713, but I am not sure whether it can fix #756 and #757. The reason is that we haven't fixed the networking layer bug in the master branch, so I have no idea whether the bug is caused by the networking layer or by the concurrency and gc modules. We need to merge #733 before fixing #756 and #757.

Issue #712 is about how we scan the table. This PR cannot address this issue, but we can have a discussion on whether we need to create a default primary index for every table, and whether the sequential scan should be executed as traversing the leaf node of primary index.

For now, this PR does not address all the issues. We are using logical pointer index management scheme, meaning that we will insert an entry into secondary index only when the corresponding key value has been updated. This brings some issues with the GC design, such as the one reported in #686. We should have a discussion on whether we should switch to physical pointer scheme.

We haven't investigated how to handle unique index.

**Comment by PauloAmora** [comment]
What exactly do you mean by unique index? The behavior when we try to create an unique index on an attribute that has duplicates?

**Comment by yingjunwu** [comment]
@PauloAmora I mean this: https://www.postgresql.org/docs/current/static/indexes-unique.html. Currently we treat unique index the same way as we treat primary index. Not sure whether it is the right way to go.

**Comment by apavlo** [comment]
@yingjunwu This is a great PR. We should have multiple reviewers look it over when you are ready.

We also need test cases to prove that you fixed the different issues.

**Comment by yingjunwu** [comment]
More test cases added. Merging this PR will fix #709 and #713.

**Comment by jarulraj** [comment]
@yingjunwu It would be great if you can add more details based on my comments, and importantly provide a high-level overview GC in a README.

**Comment by yingjunwu** [comment]
Hey @jarulraj, sorry but where I can find your comments?

File: src/include/concurrency/transaction_manager.h [hunk 1]
```diff
@@ -150,24 +150,11 @@ class TransactionManager {
...
 
-  void SetDirtyRange(std::pair<cid_t, cid_t> dirty_range) {
-    this->dirty_range_ = dirty_range;
-  }
-
   IsolationLevelType GetIsolationLevel() {
     return isolation_level_;
   }
 
  protected:
```
Comment (yingjunwu) [inline src/include/concurrency/transaction_manager.h]: This "dirty range" is used by the WBL. As the logging components are under refactoring, I removed this code for now. \cc @PauloAmora

File: test/gc/transaction_level_gc_manager_test.cpp [hunk 2]
```diff
@@ -194,12 +203,171 @@ TEST_F(TransactionLevelGCManagerTests, GCTest) {
   table.release();
 
   // DROP!
-  TestingExecutorUtil::DeleteDatabase("DATABASE");
+  TestingExecutorUtil::DeleteDatabase("DATABASE0");
```
Comment (yingjunwu) [inline test/gc/transaction_level_gc_manager_test.cpp]: Sure, but the `DeleteDatabase()` function is simply to drop the tables.

File: test/type/types_test.cpp [hunk 3]
```diff
@@ -695,5 +695,62 @@ TEST_F(TypesTests, ConflictAvoidanceTypeTest) {
                peloton::Exception);
 }
 
+
+
+TEST_F(TypesTests, RWTypeTest) {
+  std::vector<RWType> list = {
+      RWType::INVALID,
```
Comment (yingjunwu) [inline test/type/types_test.cpp]: added comments at `types.h`.

**Comment by yingjunwu** [comment]
Design docs can be found here: https://docs.google.com/document/d/1PdixbQ7WV21OnhsRkK3Oo-aq1qOrmQThLSUsfpuJ5sk/edit?usp=sharing

This PR can be merged as it is independent from any other PRs.

**Pull request merged**